### ISSUE TITLE
Support MacOS ad hoc signing on all platforms

### ISSUE THREAD

**Issue by JeyJeyGao** [dotnet/sdk#34917]
### Description

I am currently working on a cross-platform binary project, [notation-azure-kv](https://github.com/Azure/notation-azure-kv), based on .NET. We have recently run into an issue where our cross-compiled binary is unable to function properly on macOS due to the absence of macOS CodeSign.

As you might know, the CodeSign process can only be executed on macOS. This makes the release procedure for cross-platform binaries more complicated, as every developer is required to construct their own macOS-specific pipeline in order to integrate the CodeSign.

Given this, **I suggest that .NET should implement macOS CodeSign in its offerings instead of call the `codesign` binary on macOS**, so the codesign can be done on any OS. By doing so, it would greatly simplify the CodeSign process and would make .NET a more appealing and efficient language for cross-platform binary products.

Moreover, if this functionality was combined with .NET's Ahead-of-Time (AOT) compilation feature, .NET would be even more robust and efficient as a Command Line Interface (CLI) language. This would streamline the development process for all .NET developers and potentially open up .NET to more use-cases in the future.

### Reproduction Steps

On an Linux machine, run:
```sh
dotnet publish \
    --configuration Release \
    --self-contained true \
    -p:PublishSingleFile=true \
    -r osx-arm64 \
    -o "$output_dir/binary-name"
```
to build an macOS binary. Then run the binary on ARM based macOS machine, the binary will be `killed` because of no codesign.

### Expected behavior

The codesign should be done in a platform independent way like Golang. https://github.com/golang/go/blob/master/src/cmd/internal/codesign/codesign.go

### Actual behavior

If the macOS binary was not built on macOS, the binary cannot run on an ARM macOS machine. The user needs to do codesign manually.

**Comment by filipnavara** [comment]
> if someone wants to go through the effort of porting the code to do ad hoc signing to C#.

https://github.com/filipnavara/CodeSign

**Comment by agocke** [comment]
@filipnavara Looks like it's MIT, mind if we integrate it? 😄

**Comment by filipnavara** [comment]
I made it MIT explicitly because I was hoping that some of the Microsoft teams adapt it instead of rolling their own versions.

**Comment by Orgl9l** [comment]
Apologies of this comment is not supposed be be here I have no idea when well to look or go:

Given that GitHub has a MacOS build chain with MacOS runners (that have Xcode tools installed), and specific instruction for   [Installing an Apple certificate on macOS runners for Xcode development](https://docs.github.com/en/actions/deployment/deploying-xcode-applications/installing-an-apple-certificate-on-macos-runners-for-xcode-development) it would be great to understand what else in missing - "for want of a nail".

There seem to be no .Net/MSFT documentation on this (that I have been able to easily find - so find myself here and other issues https://github.com/dotnet/sdk/issues/34514, https://github.com/dotnet/runtime/pull/53913 and am not sure of this is the correct pace to post; but figure that given:

- It seems that locally built/published .Net executable single file runs on my M1 Mac fine, that:
  - Same should occur on the GitHub runner on basis that the above linked GitHub instruction are followed.

However of note, the clear omission from the above linked instructions states that:

> Note: For iOS build targets, your provisioning profile should have the extension .mobileprovision. For macOS build targets, the extension should be .provisionprofile. The example workflow above should be updated to reflect your target platform.
 
while one can (perhaps incorrectly) interpolate this change:
```
PP_PATH=$RUNNER_TEMP/build_pp.provisionprofile
```
vs:
```
PP_PATH=$RUNNER_TEMP/build_pp.mobileprovision
```
it does not cover what changes need to be here:
```
# apply provisioning profile
          mkdir -p ~/Library/MobileDevice/Provisioning\ Profiles
          cp $PP_PATH ~/Library/MobileDevice/Provisioning\ Profiles
```

Does that `dotnet publish` signing only work on an M1 runner and **_not_** the Mac Intel runner...?

Any guidance from the community would be hugely appreciated!

**Comment by agocke** [comment]
@jtschuster Anything left here?

**Comment by jtschuster** [comment]
Yes, https://github.com/dotnet/runtime/pull/110417 needs to go in, and there might be a corresponding change in sdk for bundles. Haven't made progress while working on other things.

**Comment by agocke** [comment]
@Orgl9l Sorry, missed your comment earlier. This issue is about enabling _ad hoc_ signing only, so useful only for local developer use. For actual signing you'll need to use Apple's tools on a Mac for bundling.

I don't believe we have any .NET-specific guidance for this. Most of us aren't experts on Apple's shipping requirements (and they seem to change fairly often).

**Comment by filipnavara** [comment]
> in preview 3 or is it abandoned?

It's not merged yet, so not in any preview release.

**Comment by jlumsden-mts** [comment]
It looks like it has made its way into preview4 from what I can tell:
https://github.com/dotnet/runtime/blob/release/10.0-preview4/src/installer/managed/Microsoft.NET.HostModel/MachO/BinaryFormat/SymbolTableCommand.cs

In the meantime (on .NET 8.0) I am using the Rust crate mentioned in a previous comment:
https://github.com/indygreg/apple-platform-rs
https://github.com/indygreg/apple-code-sign-action (GitHub Action variant)

**Comment by agocke** [comment]
This has been completed and should be in .NET 10.
